Add initial Cypress tests for pivot table (ref: #14049)

nemanjaglumac opened the pull request:
### Status
READY

### What does this PR accomplish?
- Adds an initial batch of Cypress tests for pivot table functionality as described in #14049

### How to test this?
- All tests should pass in CI, because no test is skipped.

### Additional notes:
- Sidebar opening doesn't have any XHR associated with it that we can wait for. I tried to rely on UI, but I'm concerned that different visualization icons and/or drag handles may not be ready and the tests may flake.
- Cypress occasionally failed for me locally saying that icon/drag handle is not visible.
- If that presents a problem, I can remove the `.should("be.visible)` and add `{force: true}` to an associated action but that is an anti-pattern.
- FWIW, so far every CI run was successful.

### Stress-testing in GitHub CI:
- I was using OSS uberjar generated from 6697758c409d5ec030ad3da456910d49d5381525
- [20/20 runs](https://github.com/nemanjaglumac/metabase-tests/actions/runs/416375653) were successful ✅

nemanjaglumac committed:
Add first pivot table Cypress test

nemanjaglumac commented on frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js:
```diff
@@ -0,0 +1,76 @@
+
+    cy.findByText("Fields to use for the table rows");
+    cy.get("@fieldOption")
+      .eq(0)
```
@paulrosenzweig please let me know if this is too specific. I could just test that these options are present, without asserting on their order.

paulrosenzweig replied:
There's probably no reason to assert the order, but I don't think it hurts.

paulrosenzweig commented on frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js:
```diff
@@ -0,0 +1,75 @@
+    );
+
+    openOrdersTable({ mode: "notebook" });
+    cy.findByText("Summarize").click();
+    cy.findByText("Count of rows").click();
+    cy.findByText("Pick a column to group by").click();
+    popover().within(() => {
+      cy.findByText(/Orders?$/).click(); // Collapse "Orders" to avoid additional scrolling
+      cy.findByText(/Users?$/).click();
+    });
+    cy.get(".ReactVirtualized__Grid").scrollTo("bottom"); // "Source" is not visible - we have to scroll
+    cy.findByText("Source").click();
+    // Add another metric
+    cy.get(".Icon-add")
+      .last() // This is fragile, but there's no other way to get to this element right now
+      .click();
+    popover().within(() => {
+      cy.findByText(/Products?$/).click();
+      cy.findByText("Category").click();
+    });
+    cy.findByText("Visualize").click();
```
It'd be nice if we could skip this UI configuration and instead push the question we want in the URL. IIRC Tom tried that once and encountered issues loading the needed frontend code in Cypress. Maybe we could write a simple helper that takes the JSON question we want and base64 encodes it for the URL.

paulrosenzweig replied:
(That's unrelated to this PR)

nemanjaglumac replied:
That would be awesome, but there's an alternative for this case.
I can use API to:
1. store this question (with `bar` visualization),
2. Open it and use UI to change the visualization to pivot table (that would still make it "ad-hoc", right?)
3. Continue asserting on UI

paulrosenzweig replied:
> that would still make it "ad-hoc", right?

Unfortunately, it wouldn't since the query is unchanged. You could add another breakout or change the aggregate to switch to ad-hoc, but that point it's probably simpler to just use the UI the whole way.

nemanjaglumac replied:
See [this line](https://github.com/metabase/metabase/pull/14051/files#diff-d5b5ec136d796378e5808beb6b14135815b5dbcc1fd7b47a06c61b7ef042d5a7R12)

paulrosenzweig replied:
I wasn't expected that to run as an ad-hoc query, but I see that it does!

I think that behavior could change in the future. We don't need to run that as ad-hoc since the display isn't part of the query. If we refine that logic to prioritize running saved cards, this test will no longer use an ad-hoc query.

paulrosenzweig commented on frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js:
```diff
@@ -0,0 +1,75 @@
+
+    cy.findByText("Visualization").click();
+    cy.get(".Icon-pivot_table").click();
+    cy.wait("@pivotTableDataset");
```
Why do we need to wait for this rather than looking for the UI?

nemanjaglumac replied:
Just as a precaution if it takes longer for calculation to finish in CI. Relying on UI leads to all sorts of issues and race conditions.
This simple test could probably go without waiting for anything, though.

nemanjaglumac replied:
No more waiting :)

paulrosenzweig commented on frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js:
```diff
@@ -0,0 +1,75 @@
+      .contains("Count");
+
+    cy.log("**-- Implicit assertions on a table itself --**");
+    cy.get(".Visualization").within(() => {
+      cy.findByText(/Users? → Source/);
+      cy.findByText(/Row totals/i);
+      cy.findByText(/Grand totals/i);
+      cy.findByText("3,520");
+      cy.findByText("4,784");
+      cy.findByText("18,760");
+    });
```
Looks good!

nemanjaglumac replied:
Thank you.
This is just the first test. I'll use this PR to group them all and then merge them to `master`.

@paulrosenzweig 
Will it make your life easier if I separate all tests in separate PRs (that will use this one as a base)?

paulrosenzweig replied:
👍 one PR works for me

nemanjaglumac committed:
Add second pivot table Cypress test

- Refactored the first quesiton
- Created helper functions to create a question and to assert on UI

nemanjaglumac commented on frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js:
```diff
@@ -0,0 +1,100 @@
+});
+
+function createPivotTableQuestion({ display = "pivot" } = {}) {
+  withSampleDataset(({ ORDERS, ORDERS_ID, PRODUCTS, PEOPLE }) => {
```
This is needed until I merge Cypress metadata extraction updates from the `release-x.37.x` branch.

nemanjaglumac replied:
[Done](https://github.com/metabase/metabase/commit/6697758c409d5ec030ad3da456910d49d5381525). Rebased and refactored this function.

nemanjaglumac commented on frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js:
```diff
@@ -0,0 +1,100 @@
+    .contains("Count");
+}
+
+function assertOnpivotTable() {
```
Note to myself - Fix case on `pivot` to `Pivot`

nemanjaglumac committed:
Add third pivot table Cypress test

nemanjaglumac committed:
Refactor tests

- Remove `withSampleDataset()`
- Visit question after creating it
- Fix case in function name

nemanjaglumac committed:
Add drag-n-drop Cypress test for pivot tables

nemanjaglumac commented on frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js:
```diff
@@ -0,0 +1,164 @@
+
+// Rely on native drag events, rather than on the coordinates
+// We have 3 "drag-handles" in this test. Their indexes are 0-based.
+function dragField(startIndex, dropIndex) {
```
@paulrosenzweig this function works fine for an initial state when we want to drag one field "on top of each other" (it gives us fail-proof pointer/location). However, if you think that we should add more drag and drop scenarios, I'd probably have to either expand it or think of some other ways to accomplish that.

nemanjaglumac commented on frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js:
```diff
@@ -0,0 +1,164 @@
+    // Give it some time to open the side-bar fully before we start dragging
+    cy.findByText(/Pivot Table options/i);
+
+    // Drag the second aggregate (Product category) from table columns to table rows
```
@paulrosenzweig 

Continuing from [this comment](https://github.com/metabase/metabase/pull/14051/files#r541291213) - are there other scenarios we should test?
- Drag row field into columns? (that gave me empty field values for the whole table)
- Assert that the `count` cannot be dragged anywhere?
- ... add your own scenario

paulrosenzweig replied:
I don't think we need to. This checks that the setting interaction works, we requery for new data, and we update the chart. Also, the UI here might change a bit still so it'd better not too have too many tests that will break.

> that gave me empty field values for the whole table
Whoops! I either need to prevent that move or fix it's display.

nemanjaglumac replied:
Ok, so this can be merged and please let me know if you think of any more test cases, or need help reproing anything related.

Thank you for reviewing this.
